### ISSUE TITLE
Async processing

### ISSUE THREAD

**Issue by Gankra** [mozilla/cargo-vet#132]
Currently the MVP was quickly hacked up with sync reqwest and no multi-threading just to keep things simple while everything is in flux. In the medium-term we should properly cleanup the code to be async and spin up multiple concurrent tasks where it makes sense.

I've already done some of that work in rust-minidump here: https://github.com/rust-minidump/rust-minidump/blob/main/breakpad-symbols/src/http.rs

Which only gets more complicated here: https://github.com/rust-minidump/rust-minidump/pull/558

Although even that is only "partially" async because the primary concurrency model for minidump-stackwalk is actually just spawning multiple instances in separate processes and them knowing how to transitionally add things to the cache. That said I have been thinking about how to make it "fully" async+multithreaded here: https://github.com/rust-minidump/rust-minidump/issues/561

I am finding myself doing a lot of duplicated logic and work between these two codebases, and it feels like there could be some very useful to factor out nugget for all the concurrent locking+caching+downloading+parsing machinery.

--------------

Things that can potentially be done concurrently in their own task:

* Each package's `suggest`
  * Want to be able to start up multiple suggests at once to queue up all the expensive subtasks below
* Each download of an imported audits.toml
  * Can we start parsing/validating it *while it's downloading*? (probably not worth it...?)
* Each fetch of a package from crates.io
  * Can we start SHAing and untarring it *while it's downloading* like we do streaming-parse in rust-minidump?
* Each diffstat of two packages (unclear if `git diff --no-index` is safe/useful to spawn multiple parallel instances of)

See https://github.com/rust-minidump/rust-minidump/issues/561 where I detail some of the machinery you would want to have to make this work.

**Comment by Gankra** [comment]
If we do want to share some code with rust-minidump on this, one thing I have been thinking about a lot is how these two codebases increasingly look like standard batch compilers, and all these various caching phases increasingly look like "incremental compilation". To that end it might be worth building on a proper incremental framework like [salsa](https://github.com/salsa-rs/salsa/) (or specifically [gluon's async fork](https://github.com/Marwes/salsa/tree/gluon_salsa). 

But we're in a kind of interesting situation where we actually have very little invalidation of our caches (we tend to cache immutable facts that should never change ever). So probably we wouldn't need the core features of this... unless we realize actually we want some invalidatable caches and then it would have been nice to have this.

**Comment by Gankra** [comment]
Oh also if we do start doing async we need to cleanup my usage of spans so that tracing knows to save+restore the spans properly whenever there's a context switch: https://docs.rs/tracing/0.1.34/tracing/struct.Span.html#in-asynchronous-code

**Comment by Gankra** [comment]
Other confounding details: possibly want a more subtle design for FileLock if we can have multiple threads in the same process concurrently modifying caches.

I think the current design is *largely* safe because our on-disk state is either "append-only cache" (fetches) or "read once at start of command, mutate in memory, write once at end of command" (diff-cache, supply-chain). 

For the "append-only" case we would already want/have in-memory synchronization so that only one task is trying to download/unpack a package. However we should still endeavour to make as many modifications "transactional" as possible with anonymous-temp-files and renames.

For the "read-once-write-once" case I think we just need to make sure we read-once *before* forking off tasks and write-once *after* joining all tasks. Essentially treating those operations as "main-thread only". Then the FileLock abstraction should do its job. Appropriate use of Send/Sync and ownership can probably statically enforce this for the writes (with incorrect reads causing a self-deadlock which is just a crash right now).

**Comment by bholley** [comment]
@Gankra Is there anything left to do here after #173?

**Comment by Gankra** [comment]
Yes this is only half-implemented. We parallelize the networking but not the filesystem i/o, and in practice the latter is actually the overwhelming bottleneck because of diffstating.

**Comment by mystor** [comment]
I think I might end up parallelizing some of the diffstat work when working on #191
